### DISCUSSION TITLE
Unexpected error when parsing a single file

### DISCUSSION THREAD

**Discussion by andylamp** [github-linguist/linguist#5660] [Classification] [answered]
I am trying to use `github-linguist` as an executable to classify individual files. Unfortunately, when a file has an external dependency `github-linguist` fails.

For example, the following passes (assuming this is named `sample1.c`):

```c
int main(int argc, char **argv)
{
  char *a = "sample_str";
  return 0;
}
```

by running:

```bash
github-linguist ./sample1.c
```

whereas the following does not,

```c
#include "asf.h"

int main(int argc, char **argv)
{
  char *a = "sample_str";
  return 0;
}
```

assuming the second extract is `sample2.c` by running:

```bash
github-linguist ./sample2.c

/home/andylamp/.rvm/gems/ruby-3.0.2/gems/github-linguist-7.17.0/lib/linguist/lazy_blob.rb:109:in `to_buffer': wrong argument type nil (expected String) (TypeError)
        from /home/andylamp/.rvm/gems/ruby-3.0.2/gems/github-linguist-7.17.0/lib/linguist/lazy_blob.rb:109:in `load_blob!'
        from /home/andylamp/.rvm/gems/ruby-3.0.2/gems/github-linguist-7.17.0/lib/linguist/lazy_blob.rb:83:in `data'
        from /home/andylamp/.rvm/gems/ruby-3.0.2/gems/github-linguist-7.17.0/lib/linguist/blob_helper.rb:127:in `binary?'
        from /home/andylamp/.rvm/gems/ruby-3.0.2/gems/github-linguist-7.17.0/lib/linguist/blob_helper.rb:155:in `text?'
        from /home/andylamp/.rvm/gems/ruby-3.0.2/gems/github-linguist-7.17.0/bin/github-linguist:98:in `github_linguist'
        from /home/andylamp/.rvm/gems/ruby-3.0.2/gems/github-linguist-7.17.0/bin/github-linguist:141:in `<top (required)>'
        from /home/andylamp/.rvm/gems/ruby-3.0.2/bin/github-linguist:23:in `load'
        from /home/andylamp/.rvm/gems/ruby-3.0.2/bin/github-linguist:23:in `<main>'

```

Note: I provided an example in `C` but the same happens with `Java` as well... (and potentially other languages that have includes)

I am using the latest version of `github-linguist` released and searched the issues but didn't find anything related to that... What is probably happening is that linguist tries to parse and expand the includes... is there a flag that I can potentially enable to skip includes...?

Thanks!

**Comment by lildude** [comment, marked as answer]
> What is probably happening is that linguist tries to parse and expand the includes...

That's most definitely not happening; Linguist isn't that clever 😁.

There's a little bit of info you're not telling us about your env or these files as the actual content works fine for me:

```console
$ github-linguist --help
Linguist v7.17.0
Detect language type and determine language breakdown for a given Git repository.

Usage: linguist <path>
       linguist <path> [--breakdown] [--json]
       linguist [--breakdown] [--json]
    -b, --breakdown                  Analyze entire repository and display detailed usage statistics
    -j, --json                       Output results as JSON
    -h, --help                       Display a short usage summary, then exit
$ /bin/cat sample1.c
int main(int argc, char **argv)
{
  char *a = "sample_str";
  return 0;
}

$ /bin/cat sample2.c
#include "asf.h"

int main(int argc, char **argv)
{
  char *a = "sample_str";
  return 0;
}

$ github-linguist ./sample1.c
./sample1.c: 6 lines (5 sloc)
  type:      Text
  mime type: text/plain
  language:  C
$ github-linguist ./sample2.c
./sample2.c: 8 lines (6 sloc)
  type:      Text
  mime type: text/plain
  language:  C

$
```

Looking at the stacktrace, this is going to be a problem loading the file content from within a git repo as it's failing here:

https://github.com/github/linguist/blob/e2f638d17598e84f6f91c2d03c2f3e27f22a2163/lib/linguist/lazy_blob.rb#L108-L110

We should only get to this point if the file is within a directory that is a git repo. 

First bit of info missing from the OP: the problem file is in a directory that is a git repo.

The failure itself indicates one of the arguments being passed to `Rugged::Blob.to_buffer()` is `nil` and not a string.

Given we've got this far to determine the file is located in a directory that is a git repo, we can be sure the `repository` argument isn't a string, which leaves the OID. The only time this can be nil is if the file in question has never been committed to the repo.

And there's the second bit of missing info: the file itself has never been committed to the repo in the directory it exists in.

And sure enough, that's exactly the case and easily reproduced:

```console
$ /bin/ls -1 sample*
sample1.c
sample2.c
$ git init foo
Initialized empty Git repository in /Users/lildude/Downloads/trash/foo/.git/
$ cp sample* foo
$ cd foo
$ git add sample1.c
$ git commit -m 'Add sample1.c'
[main (root-commit) 0ed520b] Add sample1.c
 1 file changed, 6 insertions(+)
 create mode 100644 sample1.c
$ github-linguist ./sample1.c
./sample1.c: 6 lines (5 sloc)
  type:      Text
  mime type: text/plain
  language:  C
$ github-linguist ./sample2.c
/Users/lildude/.asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/github-linguist-7.17.0/lib/linguist/lazy_blob.rb:109:in `to_buffer': wrong argument type nil (expected String) (TypeError)
	from /Users/lildude/.asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/github-linguist-7.17.0/lib/linguist/lazy_blob.rb:109:in `load_blob!'
	from /Users/lildude/.asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/github-linguist-7.17.0/lib/linguist/lazy_blob.rb:83:in `data'
	from /Users/lildude/.asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/github-linguist-7.17.0/lib/linguist/blob_helper.rb:127:in `binary?'
	from /Users/lildude/.asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/github-linguist-7.17.0/lib/linguist/blob_helper.rb:155:in `text?'
	from /Users/lildude/.asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/github-linguist-7.17.0/bin/github-linguist:98:in `github_linguist'
	from /Users/lildude/.asdf/installs/ruby/3.0.2/lib/ruby/gems/3.0.0/gems/github-linguist-7.17.0/bin/github-linguist:141:in `<top (required)>'
	from /Users/lildude/.asdf/installs/ruby/3.0.2/bin/github-linguist:23:in `load'
	from /Users/lildude/.asdf/installs/ruby/3.0.2/bin/github-linguist:23:in `<main>'
$ git add sample2.c
$ git commit -m 'Add sample2.c'
[main 42c15d4] Add sample2.c
 1 file changed, 8 insertions(+)
 create mode 100644 sample2.c
$ github-linguist ./sample2.c
./sample2.c: 8 lines (6 sloc)
  type:      Text
  mime type: text/plain
  language:  C
$
```

As you can see, adding the file to the repo solves the problem.

This is kind of expected as Linguist is primarily designed to work on git repos with added functionality to try work with individual files. 

In the case of individual files, Linguist will attempt to determine if the file is within a repo so things like overrides from the `.gitattributes` file can be applied too here:

https://github.com/github/linguist/blob/e2f638d17598e84f6f91c2d03c2f3e27f22a2163/bin/github-linguist#L86-L96

This sets things up to treat the file as part of the repo (line 93), but as you've never added a file with this name to the repo, it fails later when we try to read the content from the git blob.

This would apply to anything that treats a file as part of the repo content, like Linguist is doing here.  For example, performing a diff on the file before it has been added to the repo won't do anything:

```console
$ # Reset our last commit which added the sample2.c file to the repo
$ git reset "HEAD^1"
$ git status
On branch main
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	sample2.c

nothing added to commit but untracked files present (use "git add" to track)
$ git diff sample2.c 

$
```

So in summary, you will need to either add the file to the repo in question or perform your analysis against the file outside of a directory that is a git repo.

**Reply by andylamp** [reply]
hurray @lildude! that's more than I could ever ask for as an explanation. Yes indeed, the files where inside a `git` repository but _not_ committed. Thank you not only for providing a reason of the problem but also giving an extremely good explanation about it.

Now that I think about it - this might make a good FAQ as am sure others would have tried the same thing (and probably got similar errors...).

Would it be worth refactoring the above and putting it there by the means of a PR?

**Reply by lildude** [reply]
Yeah, that would be a good idea. I've been meaning to resurrect https://github.com/github/linguist/pull/4271 for a while but haven't had the time and this could go with it.
